### PR TITLE
Configure PsysPL2, PsysPL3, PL4 according to charger limits

### PR DESCRIPTION
**Pull request by mkopec**

### CODE REVIEW

File: src/board/system76/common/power.c [hunk 1]
```diff
@@ -334,13 +334,42 @@ void power_off(void) {
...
+        uint16_t res;
+        do {
+            res = peci_wr_pkg_config(PECI_REG_PKG_CFG_PSYS_PL2, 0, PECI_PSYS_PL2(watts));
+            delay_us(1);
+        } while (res != 0x40 && retry--);
+        DEBUG("PECI SET PsysPL2 %ld = %d\n", watts, res);
+
+        retry = 10;
+        do {
+            res = peci_wr_pkg_config(PECI_REG_PKG_CFG_PL3, 0, PECI_PL3(watts, 6, 4));
```
Comment (miczyg1) [inline src/board/system76/common/power.c:361]: PL3 should be 2W higher than PL2 per Intel recommendations

File: src/board/system76/common/power.c [hunk 2]
```diff
@@ -357,14 +386,9 @@ void power_cpu_reset(void) {
     fan_reset();
     // Reset KBC and touchpad states
     kbled_reset();
-    // Set PL4
+    // Set PsysPL2
```
Comment (miczyg1) [inline src/board/system76/common/power.c:389]:
```suggestion
    // Set power limits
```

File: src/board/system76/common/power.c [hunk 3]
```diff
@@ -357,14 +386,9 @@ void power_cpu_reset(void) {
...
     //TODO: if this returns false, retry?
-    power_peci_limit(
-        // AC is connected
-        (!gpio_get(&ACIN_N)) &&
-        // There is available current
-        (battery_charger_input_current >= CHARGER_INPUT_CURRENT)
-    );
+    power_peci_limit(gpio_get(&ACIN_N));
```
Comment (miczyg1) [inline src/board/system76/common/power.c:391]: If this is done right after CPU reset, BIOS will not override those values? Or values set by PECI are stored elsewhere and take precedence?

File: src/board/system76/common/power.c [hunk 4]
```diff
@@ -334,13 +334,42 @@ void power_off(void) {
...
+        retry = 10;
+        do {
+            res = peci_wr_pkg_config(PECI_REG_PKG_CFG_PL3, 0, PECI_PL3(watts, 6, 4));
+            delay_us(1);
+        } while (res != 0x40 && retry--);
+        DEBUG("PECI SET PL3 %ld = %d\n", watts, res);
+
+        retry = 10;
+        do {
+            res = peci_wr_pkg_config(PECI_REG_PKG_CFG_PL4, 0, PECI_PL4(watts));
```
Comment (miczyg1) [inline src/board/system76/common/power.c:368]: Not sure if PL4 should be the same as PL2/PL3. PL4 is Potential Pewak Power of the SoC, but what about the ROP (Rest of Platform)? I think it should be slightly lower than psysPL2 (say 5W or 10W at least)

File: src/board/system76/common/usbpd/tps65987.c [hunk 5]
```diff
@@ -25,6 +25,9 @@ static int16_t usbpd_current_limit(void) {
         if (value[0] == 6) {
             uint32_t pdo = ((uint32_t)value[1]) | (((uint32_t)value[2]) << 8) |
                 (((uint32_t)value[3]) << 16) | (((uint32_t)value[4]) << 24);
+            if (!pdo) {
+                return -0x2000;
```
Comment (miczyg1) [inline src/board/system76/common/usbpd/tps65987.c:29]: What does this value mean?

Also no need braces for single statements

File: src/board/system76/common/usbpd/tps65987.c [hunk 6]
```diff
@@ -113,6 +118,10 @@ void usbpd_event(void) {
         update = true;
 
         DEBUG("SINK_CTRL %d\n", sink_ctrl);
+
+        if (sink_ctrl) {
+            wait_for_contract = 100;
+        }
```
Comment (miczyg1) [inline src/board/system76/common/usbpd/tps65987.c:124]:
```suggestion
        if (sink_ctrl)
            wait_for_contract = 100;
```

File: src/board/system76/common/usbpd/tps65987.c [hunk 7]
```diff
@@ -122,25 +131,39 @@ void usbpd_event(void) {
         update = true;
     }
 
+    if (wait_for_contract) {
+        update = true;
+    }
```
Comment (miczyg1) [inline src/board/system76/common/usbpd/tps65987.c:136]:
```suggestion
    if (wait_for_contract)
        update = true;
```

**Review by miczyg1 — CHANGES_REQUESTED** [review]

File: src/board/system76/common/power.c (hunk 1 above)
Comment (mkopec) [reply src/board/system76/common/power.c:361]: PL3 and PsysPL3 share the same register, if Psys is enabled then this register is used for PsysPL3 and not PL3 (doc #613972)

File: src/board/system76/common/power.c (hunk 3 above)
Comment (mkopec) [reply src/board/system76/common/power.c:391]: The BIOS will override these values. Ideally we could do what some chromebooks do and have the same PsysPLx calculation in coreboot as well.

File: src/board/system76/common/power.c (hunk 4 above)
Comment (mkopec) [reply src/board/system76/common/power.c:368]: To be honest I duplicated chromebook behavior here. But yes, PL4 can be set lower, perhaps to `POWER_LIMIT_AC` when on AC power.

File: src/board/system76/common/usbpd/tps65987.c (hunk 7 above)
Comment (mkopec) [reply src/board/system76/common/usbpd/tps65987.c]: done

File: src/board/system76/common/usbpd/tps65987.c (hunk 6 above)
Comment (mkopec) [reply src/board/system76/common/usbpd/tps65987.c]: done

File: src/board/system76/common/usbpd/tps65987.c (hunk 5 above)
Comment (mkopec) [reply src/board/system76/common/usbpd/tps65987.c]: No valid PDO was returned.

(Actually -0x2000 is already used below on i2c read failure, so changed this to 0x1000 instead.)

File: src/board/system76/common/power.c (hunk 2 above)
Comment (mkopec) [reply src/board/system76/common/power.c]: done

File: src/board/system76/common/power.c (hunk 1 above)
Comment (miczyg1) [reply src/board/system76/common/power.c:361]: Yes, it is recommended as:

PL3 = PL2 + 2W
PsysPL3 = PsysPL2 + 2W

File: src/board/system76/common/power.c (hunk 3 above)
Comment (miczyg1) [reply src/board/system76/common/power.c:391]: Generally BIOS should set the static maximum values for those power limits, so EC shouldn't set higher values than BIOS does. But if we end up in situation when we have to lower the power limits, then it should be called in some event handling function.

It is fine to set those values right after PCU reset if and only if the values are pretty much the same as the BIOS settings.

File: src/board/system76/common/power.c (hunk 4 above)
Comment (miczyg1) [reply src/board/system76/common/power.c:368]: > ROP is included in the Psys measurements and is handled by Psys limits.

True, but the thing is PL4 does not account for ROP.

File: src/board/system76/common/usbpd/tps65987.c (hunk 5 above)
Comment (miczyg1) [reply src/board/system76/common/usbpd/tps65987.c]: Ughh. Can we add some error codes in macros for it?

File: src/board/system76/common/power.c (hunk 1 above)
Comment (mkopec) [reply src/board/system76/common/power.c]: done

File: src/board/system76/common/power.c (hunk 4 above)
Comment (mkopec) [reply src/board/system76/common/power.c]: Okay, set PL4 to POWER_LIMIT_AC, when power supply capacity allows.

File: src/board/system76/common/usbpd/tps65987.c (hunk 5 above)
Comment (mkopec) [reply src/board/system76/common/usbpd/tps65987.c]: Added an enum for error codes.

File: src/board/system76/common/power.c (hunk 3 above)
Comment (miczyg1) [reply src/board/system76/common/power.c:394]: > Hmm, the problem here is that setting the limits in coreboot is what enables the Psys functionality on the SoC side at all.

That's also right.

> I would prefer to only set them in the EC but that doesn't work.

That si why EC should set it asynchronically on some event function. Let's leave it like that, we can still fix things after some more testing will be done.

**Review by miczyg1 — APPROVED** [review]

**Comment by miczyg1** [comment]
All we need right now is for the NV4x TGL build to not fail

File: src/board/system76/common/usbpd/tps65987.c [hunk 8]
```diff
@@ -18,53 +18,59 @@ void usbpd_init(void) {
...
+    USBPD_ERR_REG_LEN = 0x3000,
+    // Unsupported PDO type was received
+    USBPD_ERR_PDO_TYPE_UNSUP = 0x4000,
+};
+
+#define PDO_KIND(pdo) ((uint8_t)((pdo) >> 30 & 0b11))
+#define PDO_KIND_FIXED 0b00
+#define PDO_KIND_BATTERY 0b01
+#define PDO_KIND_VARIABLE 0b10
+#define PDO_KIND_AUGUMENTED 0b11
```
[the remark this answers is not in the document]
Comment (mkopec) [reply src/board/system76/common/usbpd/tps65987.c]: Fixed in https://github.com/Dasharo/ec/pull/40/commits/1475f5d128b1e2dc42cb5237626d4469307c665a

**Review by miczyg1 — CHANGES_REQUESTED** [review]

File: src/board/system76/common/usbpd/tps65987.c [hunk 9]
```diff
@@ -18,53 +18,62 @@ void usbpd_init(void) {
...
+};
+
+#define PDO_KIND(pdo) ((uint8_t)((pdo) >> 30 & 0b11))
+
+enum {
+    PDO_KIND_FIXED = 0b00,
+    PDO_KIND_BATTERY = 0b01,
+    PDO_KIND_VARIABLE = 0b10,
+    PDO_KIND_AUGUMENTED = 0b11,
+};
```
[the remark this answers is not in the document]
Comment (miczyg1) [reply src/board/system76/common/usbpd/tps65987.c:39]: @mkopec could you apply and test?

[the remark this answers is not in the document]
Comment (mkopec) [reply src/board/system76/common/usbpd/tps65987.c]: Applied in https://github.com/Dasharo/ec/pull/40/commits/96647d735d49a47be63e7421729ba0b51635a425 , works correctly

**Review by miczyg1 — APPROVED** [review]

**Comment by miczyg1** [comment]
@mkopec I have been waiting for the last check to pass :)

**Pull request merged**